### DISCUSSION TITLE
request license token failed (104)

### DISCUSSION THREAD

**Discussion by AkihikoWada** [dashingsoft/pyarmor#1409] [Q&A] [answered]
I tried initial registration, but failed in the procedure.
Attaching the log as below, could you help me to solve the problem, please?
I would like to know the cause of this error.
For your information, I edit registration number and the application name for privacy.

D:\Users\0000011110596\Downloads\Pyarmor>pyarmor reg -p "App Name" pyarmor-regcode-xxxx.txt
INFO     Python 3.10.2
INFO     Pyarmor 8.3.0 (trial), 000000, non-profits
INFO     Platform windows.x86_64
INFO     prepare "pyarmor-regcode-xxxx.txt"
INFO     query key file from server
This license registration information will be

License Type    : pyarmor-basic
License To      : xxxxxx
License Product : App Name

This license is about to be used for product "App Name"

Are you sure to continue? (yes/no) y
INFO     process activation file "pyarmor-regcode-xxxx.txt"
INFO     send request to server
INFO     write registration file "pyarmor-regfile-xxxx.zip"
INFO     register "pyarmor-regfile-xxxx.zip"
INFO     update license token
ERROR    request license token faild (104)

**Comment by jondy** [comment, marked as answer]
You may block server `pyarmor.dashingsoft.com`

**Comment by jondy** [comment]
First make sure no block `pyarmor.dashingsoft.com`.

There is also request limitations per seconds, do not run `pyarmor reg` too many times in one minute. Especially don't run `pyarmor reg` in loop by shell script.

I also increase the request limitations in server side, but too many requests in one seconds still may be rejected.

**Comment by jondy** [comment]
make sure firewall doesn't block the response of license server. If possible, turn off the firewall to verify it.

In Windows make sure ``pytransform3.pyd`` is not blocked, in other platforms make sure ``pytransform3.so`` is not blocked. This binary file need connect to `pyarmor.dashingsoft.com` to request token for online obfuscation.

**Comment by AkihikoWada** [comment]
@jondy 

Thank you for your advice.
Configuring to allow the communication to pyarmor.dashingsoft.com, I can finish initial registration properly.

By the way, can I ask which number of the port is used to communicate to the server?
I released all the restriction of the network this time, but it is a little dangerous.

**Comment by jondy** [comment]
Only `pyarmor.dashingsoft.com` and port `80` is used by `pytransform3.pyd` in Windows.

**Comment by takahiroinagakiTT** [comment]
I am having the same problem, so let me ask here for a solution. i understand that i need to allow pyarmor.dashingsoft.com and port 80 to resolve ERROR request license token faild (104). However, in order to get these permissions I need to know what I am sending/receiving during activation. Do you upload files during activation? or not? What information is being sent and received at that time?

**Comment by jondy** [comment]
Please check [Pyarmor EULA](https://github.com/dashingsoft/pyarmor/blob/master/LICENSE) item 3
```
3.1 This software does nothing except the features described in the software
    documentation.

3.2 Using Pyarmor Basic and Pro License, this software need internet
    connection to request authorization. And only the related files of
    Pyarmor License, serial number of hard disk, Ethernet address, IPv4/IPv6
    address, hostname will be sent to remote server for verification.

3.3 Except 3.2, this software does not record and collect any other device
    information, and need not connect to internet.
```

**Comment by AkihikoWada** [comment]
@jondy 
Thank you for your time and support.
Could you please provide us with an advice little more.

> In Windows make sure pytransform3.pyd is not blocked,

What specifically should we check if pytransform3.pyd is blocked or not?

**Comment by jondy** [comment]
I have no more expirences on Windows Defender, please check Windows Defender guide about how to check its log.

And someone else reports http proxy may not work too, http proxy will be supported in next release v8.3.6

**Comment by chickahoona** [comment]
@jondy I think its not really fair to always point people to a local firewall. That error in particular occurs across multiple devices / networks and so on spontaniously from time to time. Fixing it always requires jsut trying again and again and to hope that it will pass. A local firewall for sure would have the same picture yet most people here have the issue because of your hoster and them not receiving the request.
I can only beg you to put your pyarmor.dashlane service up on a different service, something that is globally reachable

**Comment by jondy** [comment]
The server works fine, but there are request limitations in one minute from same IP, run `pyarmor reg` no more than 3 times in 1 minute.

**Reply by chickahoona** [reply]
Jondy. Thats not the case. I run maybe 1 request every 5 minutes, so far away from any limit. Its just that things time out. I believe you that your server works, but the provider with its network doesn't. So some requests just don't reach your server.

**Reply by chickahoona** [reply]
![image](https://github.com/dashingsoft/pyarmor/assets/8075146/99ef3b21-381c-45fa-8797-89f5834376f4)
You already see where things go wrong. So if you want to avoid this timeout story you will have to move your infrastructure to a provider that is not alibaba, or (I am not sure if they offer that) move the servers to a location in US or Europe, or setup "mirrored" systems there and route requests accoridingly.

**Comment by jondy** [comment]
Got it, thanks. I'll first ask for Alibaba Cloud Team to find solutions.

**Comment by jondy** [comment]
Just for test, add one line to `/etc/hosts/`
```
8.218.217.13 pyarmor.dashingsoft.com
```

And try it again. `8.218.217.13` is a proxy host in HongKong.

**Reply by chickahoona** [reply]
I'd love to help you out here, yet my build infrastructure is using docker containers that run on dynamically provisioned VMs on hetzner, so I am not sure how to accomplish that. Docker for example will usually ignore /etc/hosts entries ... I don't have a DNS server there that I could use ... :(

**Comment by jondy** [comment]
Using option `--add-host` for `docker run`, for example
```
docker run --add-host pyarmor.dashingsoft.com:8.218.217.13 ...
```

**Reply by chickahoona** [reply]
Just an update on this, that I have updated things and are testing it but so far its MUCH more reliable.

**Comment by jondy** [comment]
Got it, thanks.
